extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elasticsearch/impl/ElasticProcessorMetaSupplier.java

```diff
@@ -0,0 +1,210 @@
+import com.hazelcast.internal.json.Json;
+import com.hazelcast.internal.json.JsonArray;
+import com.hazelcast.internal.json.JsonObject;
+import com.hazelcast.internal.json.JsonValue;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.elasticsearch.ElasticsearchSourceBuilder;
+import com.hazelcast.jet.elasticsearch.impl.Shard.Prirep;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.Vertex;
+import com.hazelcast.jet.core.processor.Processors;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.partition.strategy.StringPartitioningStrategy;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestHighLevelClient;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.nCopies;
+import static java.util.Optional.empty;
+import static java.util.logging.Level.FINER;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toList;
+
+public class ElasticProcessorMetaSupplier<T> implements ProcessorMetaSupplier {
+
+    private static final long serialVersionUID = 1L;
+
+    private transient ILogger logger;
+
+    @Nonnull
+    private final ElasticsearchSourceBuilder<T> builder;
+
+    private Map<String, List<Shard>> assignedShards;
+    private transient Address ownerAddress;
+
+    public ElasticProcessorMetaSupplier(@Nonnull ElasticsearchSourceBuilder<T> builder) {
+        this.builder = builder;
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, String> getTags() {
+        return emptyMap();
+    }
+
+    @Override
+    public int preferredLocalParallelism() {
+        if (builder.slicing() || builder.coLocatedReading()) {
+            return Vertex.LOCAL_PARALLELISM_USE_DEFAULT;
+        } else {
+            return 1;
+        }
+    }
+
+
+    @Override
+    public void init(@Nonnull Context context) throws Exception {
+        logger = context.logger();
+
+        List<Shard> shards = readShards();
+        if (builder.coLocatedReading()) {
+            List<String> addresses = context
+                    .jetInstance().getCluster().getMembers().stream()
+                    .map(m -> uncheckCall((() -> m.getAddress().getInetAddress().getHostAddress())))
+                    .collect(toList());
```

Comment:
by using `toSet()` here we can remove `new HashSet<>(addresses)` in `assignShards` method